oauth-common/src/main/java/io/strimzi/kafka/oauth/common/HttpUtil.java

```diff
@@ -0,0 +1,121 @@
+    }
+
+    public static <T> T post(URI uri, String authorization, String contentType, String body, Class<T> responseType) throws IOException {
+        return postOrGet(uri, null, null, authorization, contentType, body, responseType);
+    }
+
+    public static <T> T post(URI uri, SSLSocketFactory socketFactory, String authorization, String contentType, String body, Class<T> responseType) throws IOException {
+        return postOrGet(uri, socketFactory, null, authorization, contentType, body, responseType);
+    }
+
+    public static <T> T post(URI uri, SSLSocketFactory socketFactory, HostnameVerifier verifier, String authorization, String contentType, String body, Class<T> responseType) throws IOException {
+        return postOrGet(uri, socketFactory, verifier, authorization, contentType, body, responseType);
+    }
+
+    public static <T> T postOrGet(URI uri, SSLSocketFactory socketFactory, HostnameVerifier hostnameVerifier, String authorization, String contentType, String body, Class<T> responseType) throws IOException {
+        HttpURLConnection con;
+        try {
+            con = (HttpURLConnection) uri.toURL().openConnection();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("Malformed token endpoint url: " + uri);
+        }
+
+        if (con instanceof HttpsURLConnection) {
+            HttpsURLConnection scon = (HttpsURLConnection) con;
+            if (socketFactory != null) {
+                scon.setSSLSocketFactory(socketFactory);
+            }
+            if (hostnameVerifier != null) {
+                scon.setHostnameVerifier(hostnameVerifier);
+            }
+        } else if (socketFactory != null) {
+            log.warn("SSL socket factory set but url scheme not https ({})", uri);
+        }
+
+        con.setUseCaches(false);
+        if (body != null) {
+            con.setDoOutput(true);
+        }
+
+        con.setRequestMethod(body != null ? "POST" : "GET");
+        if (authorization != null) {
+            con.setRequestProperty("Authorization", authorization);
+        }
+        con.setRequestProperty("Accept", "application/json");
+        if (body != null && body.length() > 0) {
+            if (contentType == null) {
+                throw new IllegalArgumentException("contentType must be set when body is not null");
+            }
+            con.setRequestProperty("Content-Type", contentType);
+        }
+
+        try {
+            con.connect();
+        } catch (ConnectException e) {
+            throw new IOException("Failed to connect to: " + uri, e);
+        }
+
+        if (body != null && body.length() > 0) {
+            OutputStream out = con.getOutputStream();
+            out.write(body.getBytes(StandardCharsets.UTF_8));
+            out.close();
+        }
+
+        int code = con.getResponseCode();
+        if (code != 200) {
+            InputStream err = con.getErrorStream();
```

Comment:
This never gets closed.

Reply:
Doh! Good catch!

Reply:
In fact this one gets closed in IOUtil.copy method. If reading from inpuststream fails it means inputstream is broken. The implementation behind it should close it itself. If it is consumed fully it should also be closed by underlying implementation itself. But to not rely on proper impl of other people's code the copy() method takes care of closing it.